server/sonar-server/src/main/java/org/sonar/server/computation/formula/AverageFormula.java

```diff
@@ -0,0 +1,178 @@
+      if (byValue > 0d) {
+        return Optional.of(Measure.newMeasureBuilder().create(mainValue / byValue));
+      }
+    }
+    return Optional.absent();
+  }
+
+  @Override
+  public String getOutputMetricKey() {
+    return outputMetricKey;
+  }
+
+  public static class Builder {
+    private String outputMetricKey;
+    private String mainMetric;
+    private String byMetric;
+    @CheckForNull
+    private String fallbackMetric;
+
+    private Builder() {
+      // prevents instantiation outside static method
+    }
+
+    public static Builder newBuilder(){
+      return new Builder();
+    }
+
+    public Builder setOutputMetricKey(String m) {
+      this.outputMetricKey = m;
+      return this;
+    }
+
+    public Builder setMainMetricKey(String m) {
+      this.mainMetric = m;
+      return this;
+    }
+
+    public Builder setByMetricKey(String m) {
+      this.byMetric = m;
+      return this;
+    }
+
+    public Builder setFallbackMetricKey(@Nullable String m) {
+      this.fallbackMetric = m;
+      return this;
+    }
+
+    public AverageFormula build() {
+      Objects.requireNonNull(outputMetricKey, "Output metric key cannot be null");
+      Objects.requireNonNull(mainMetric, "Main metric Key cannot be null");
+      Objects.requireNonNull(byMetric, "By metric Key cannot be null");
+      return new AverageFormula(this);
+    }
+  }
+
+  class AverageCounter implements Counter<AverageCounter> {
+
+    private boolean initialized = false;
+
+    private double mainValue = 0d;
+    private double byValue = 0d;
+
+    @Override
+    public void aggregate(AverageCounter counter) {
+      addValuesIfPresent(counter.getMainValue(), counter.getByValue());
+    }
+
+    @Override
+    public void aggregate(CounterContext counterContext) {
+      Optional<Double> mainValueOptional = getDoubleValue(counterContext.getMeasure(mainMetric));
+      if (!mainValueOptional.isPresent() && fallbackMetric != null) {
+        mainValueOptional = getDoubleValue(counterContext.getMeasure(fallbackMetric));
+      }
+      Optional<Double> byValueOptional = getDoubleValue(counterContext.getMeasure(byMetric));
+      addValuesIfPresent(mainValueOptional, byValueOptional);
+    }
+
+    private void addValuesIfPresent(Optional<Double> counterMainValue, Optional<Double> counterByValue) {
+      if (counterMainValue.isPresent() && counterByValue.isPresent()) {
+        initialized = true;
+        mainValue += counterMainValue.get();
+        byValue += counterByValue.get();
+      }
+    }
+
+    public Optional<Double> getMainValue() {
+      return getValue(mainValue);
+    }
+
+    public Optional<Double> getByValue() {
+      return getValue(byValue);
+    }
+
+    private Optional<Double> getValue(double value) {
+      if (initialized) {
+        return Optional.of(value);
+      }
+      return Optional.absent();
+    }
+  }
+
+  private static Optional<Double> getDoubleValue(Optional<Measure> measureOptional) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "getDoubleValue" is 12 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)